File: app/src/main/java/life/plank/juna/zone/view/adapter/board/creation/BoardColorThemeAdapter.java
package life.plank.juna.zone.view.adapter.board.creation;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.lang.ref.WeakReference;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.hdodenhof.circleimageview.CircleImageView;
import life.plank.juna.zone.R;
import life.plank.juna.zone.util.BoardColorFactory;

public class BoardColorThemeAdapter extends RecyclerView.Adapter<BoardColorThemeAdapter.BoardColorThemeViewHolder> {

    private static int selectedIndex = -1;
    private List<String> boardColorList;

    public BoardColorThemeAdapter() {
        this.boardColorList = BoardColorFactory.getAllColors();
    }

    @Override
    public BoardColorThemeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new BoardColorThemeViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(R.layout.item_board_color_theme, parent, false),
                this
        );
    }

    @Override
    public void onBindViewHolder(BoardColorThemeViewHolder holder, int position) {
        holder.colorThemeImageView.setImageDrawable(new ColorDrawable(Color.parseColor(boardColorList.get(position))));
        holder.imageSelectionMarker.setVisibility(selectedIndex == position ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public int getItemCount() {
        return boardColorList.size();
    }

    public String getSelectedColor() {
        if (selectedIndex >= 0)
            return boardColorList.get(selectedIndex);
        return null;
    }

    static class BoardColorThemeViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.color_theme_image_view)
        CircleImageView colorThemeImageView;
        @BindView(R.id.image_selection_marker)
        ImageView imageSelectionMarker;
        private WeakReference<BoardColorThemeAdapter> ref;

        BoardColorThemeViewHolder(View itemView, BoardColorThemeAdapter adapter) {
            super(itemView);
            ref = new WeakReference<>(adapter);
            ButterKnife.bind(this, itemView);
        }

        @OnClick(R.id.root_layout)
        public void onColorThemeSelected() {
            int previousSelection = selectedIndex;
            imageSelectionMarker.setVisibility(selectedIndex == getAdapterPosition() ? View.INVISIBLE : View.VISIBLE);
            selectedIndex = getAdapterPosition();
            ref.get().notifyItemChanged(previousSelection);
            ref.get().notifyItemChanged(selectedIndex);
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/view/adapter/board/creation/BoardColorThemeAdapter.kt
package life.plank.juna.zone.view.adapter.board.creation

import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.item_board_color_theme.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.util.facilis.onDebouncingClick
import life.plank.juna.zone.util.view.BoardColorFactory

class BoardColorThemeAdapter : RecyclerView.Adapter<BoardColorThemeAdapter.BoardColorThemeViewHolder>() {

    private val boardColorList: List<String> = BoardColorFactory.getAllColors()

    val selectedColor: String?
        get() = if (selectedIndex >= 0) boardColorList[selectedIndex] else null

    companion object {
        private var selectedIndex = -1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BoardColorThemeViewHolder =
            BoardColorThemeViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_board_color_theme, parent, false))

    override fun onBindViewHolder(holder: BoardColorThemeViewHolder, position: Int) {
        holder.itemView.run {
            color_theme_image_view.setImageDrawable(ColorDrawable(Color.parseColor(boardColorList[position])))
            image_selection_marker.visibility = if (selectedIndex == position) View.VISIBLE else View.INVISIBLE

            onDebouncingClick {
                val previousSelection = selectedIndex
                image_selection_marker.visibility = if (selectedIndex == position) View.INVISIBLE else View.VISIBLE
                selectedIndex = position
                notifyItemChanged(previousSelection)
                notifyItemChanged(selectedIndex)
            }
        }
    }

    override fun getItemCount(): Int = boardColorList.size

    class BoardColorThemeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}